### PR TITLE
Add git branches self-guided lesson

### PR DESCRIPTION
**Pull request by illicitonion**
## What does this change?

Module: Git
Week(s): Branches

Git branches lesson, based on discussion in https://docs.google.com/document/d/1IXYc4HZ0lQA5UuzTTOuf3zbDPY8lFOj4zT5exuE_BQ0/edit

This is designed for self-study as coursework, but the exercises are in-line in the lesson, which makes it not quite fit in with the sidebar style of the rest of the module.

There are two open TODOs:
1. Create a few diagrams (I can give this a go if no one more artistically minded has time)
2. Record a demo video of a PR creation/merge/branch-deletion cycle. I can give this a go, too, though I've never screen-casted before so would gladly defer to someone with more experience :)

[Rendered version](https://github.com/CodeYourFuture/syllabus/blob/git-branches/docs/git/branches.md)

## Who needs to know about this?

@ChrisOwen101 @timhamrouge @40thieves

### CODE REVIEW

File: docs/git/branches.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+TODO: Image of two divergent branches with shared history, highlighting which is shared and which is divergent.
+
+### Why would we want to do this?
+
+There are a few reasons. The main one is: So you can work on more than one thing at the same time.
+
+Let's look at our repo of recipes. Right now, it contains one recipe, for a brownie in a mug. There are a couple of changes we want to make:
+
+1. We want to add an optional topping to put on top of the brownie (maybe ice cream, maybe you have a better idea). We'll need to add it both as an extra ingredient, and an extra step at the end.
+2. A lot of recipes start with a story of how the author found them, or some time they've really enjoyed the recipe. We want to add a story, before the "What you need" section.
```
Comment (alismith-dev) [inline docs/git/branches.md]: \beginsnark
Can we make sure it's a really unnecessarily long story that takes ages to scroll through before you get to the thing you actually wanted to read?
\endsnark

EDIT: I'm just going to mark this as resolved immediately. Apologies, I couldn't resist the temptation of the joke 😂

File: docs/git/branches.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+`--create` - we want git to create the branch we're switching to.
+
+`story` - the branch we want to switch to (and create) is called `story`.
+
+Running this command doesn't change any of the files in our repository - they stay the same, but if we commit changes, they will only get committed on this branch. If we switched back to the `main` branch, the changes would be un-done. But if we switch back to `story`, they'll come back!
+
+### How do I know what branch I'm on?
+
+You've already run one command which tells you this, you just may not have noticed!
```
Comment (alismith-dev) [inline docs/git/branches.md]: Have we already run this?

File: docs/git/branches.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+When you've written a bit, make a commit (make sure to give it a good commit message!)
+
+Explore the history with `git log` - make sure you understand the commits in your history and why they're there.
+
+### Switching back to main
+
+Can you guess how to switch back to the `main` branch?
+
+What do you think will happen when you do?
```
Comment (alismith-dev) [inline docs/git/branches.md]: Sorry if this is overly nitpicky for this level of review, but I'd make sure the emphasis is on what happens to your *files* when you switch.

File: docs/git/branches.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+### Exercise 2 (5 minutes)
+
+Switch back to the `main` branch (hint: The command to run is like the one you used to create a new branch, but you're not creating a new one).
+
+Look at the contents of `mug-brownie.md` - what's changed?
+
+Explore the history with `git log` - make sure you understand the commits in your history and why they're there.
+
+Share with the class what you saw.
```
Comment (alismith-dev) [inline docs/git/branches.md]: In the past I've found stuff like this a bit unstructured. Discussing with a buddy/small group (and maybe reporting back to the larger group) works better.

File: docs/git/branches.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+Make another branch based on `main` called `extra-topping`.
+
+Add a new topping to the recipe. Make sure to include it both in the ingredients list, and the steps.
+
+Commit your changes.
+
+### Managing branches for pull requests
+
+TODO: Talk through pushing branches to remotes, and making pull requests from them
```
Comment (alismith-dev) [inline docs/git/branches.md]: Might be a good idea to discuss pushing `main` to a remote first? They wouldn't be familiar with that yet.

File: docs/git/branches.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+### Teacher-led example
+
+TODO: Make a pull request from one of your branches. Merge it. Delete the branch. Explain that it's ok to delete the branch because we know the changes are safely on `main`.
```
Comment (alismith-dev) [inline docs/git/branches.md]: It's not the best formatting of all time but `git log --graph` might be useful here?

**Review by alismith-dev — COMMENTED** [review]
Looks good! I like the overall structure.

To answer the open questions, the only one I would be tempted to include is `git diff`. IIRC we do teach them to check the diff in Github Desktop before pushing. Quickly showing how to do this via CLI is probably a good idea. The main drawback as I see is that you end up in a `less` mode, which isn't super intuitive.

File: docs/git/branches.md (hunk 2 above)
Comment (illicitonion) [reply docs/git/branches.md]: I'm hoping `git status` will come up in the intro to git cli class, but may need to edit this if not :)

File: docs/git/branches/branches.md [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+It would be better if we could send out one pull request for the ice cream change, and another for the story. We can do this - with branches:
+
+![Committing to another Branch](./assets/git-example-commit.gif)
+
+We can then add new commits on top of each branch, separately, and later on merge both branches to main.
+
+### How do we create a new branch?
+
+```console
+$ git switch --create story
```
Comment (ChrisOwen101) [inline docs/git/branches/branches.md:67]: Good call with going with `switch` . I always felt having to create + checkout across two steps was a stumbling block for learners

File: docs/git/branches/branches.md [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+$ git branch
+  main
+* story
+```
+
+`git branch` tells you what branches exist in your repository, and puts a star next to the one you're currently switched to.
+
+### Exercise 1 (15 minutes)
+
+:::tip Exercise
```
Comment (ChrisOwen101) [inline docs/git/branches/branches.md:109]: I've added fancy callouts for the exercises - again, welcome to remove if you think it's distracting

File: docs/git/branches/branches.md [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+### Teacher-led example / video
+
+TODO: Make a pull request from one of your branches. Merge it. Delete the branch. Explain that it's ok to delete the branch because we know the changes are safely on `main`.
+TODO: Show `git status` showing that branches are out of sync.
+
+### Exercise 4 (15 minutes)
+
+Make pull requests for both of your branches (one with a story, one with an extra ingredient).
+
+Pair up with another trainee in your class. Get a them to review each one of your two pull requests. You should also review theirs.
```
Comment (ChrisOwen101) [inline docs/git/branches/branches.md]: Maybe pull this task out to a piece of coursework? 

I've found that students put more importance on it if it's in Google Classroom as a task - I think students may skip this step unless they're already pairing

**Review by ChrisOwen101 — COMMENTED** [review]
Great work!

Happy for this to be merged whenever you see fit :)

File: docs/git/branches/branches.md [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+- Create a branch, and add commits to it.
+- Switch between branches.
+- Create a pull request from a branch to the main branch.
+- Work with another trainee on non-conflicting changes to a single branch.
+
+## 1) What is a branch?
+
+We saw in the first Git lesson that Git stores the history of your changes as a list of commits, one after the other:
+
+![Committing to the Master Branch](./assets/git-main-commit.gif)
```
[the remark this answers is not in the document]
Comment (illicitonion) [reply docs/git/branches/branches.md:37]: Really nice! It's a shame they don't support `git switch` :( I'll add a note to the first one explaining why it looks different than what we're teaching :)

**Comment by illicitonion** [comment]
Thanks for the feedback (and diagrams)! I've acted on the suggestions and tweaked a bit, and merged - very happy to review any improvements people have!

**Pull request merged**